### PR TITLE
fix: migration for application export clean up

### PR DESCRIPTION
**Pull request by YazeedLoonat**
## Issue Overview

This PR addresses [#issue](https://github.com/bloom-housing/bloom/issues/3007)

- [x] This change addresses only certain aspects of the issue

## Description
This covers the migration for updating application data for programs and preferences. What we were running into was that old data (program/preference data) and new data (multiselect question data) existing on applications for a listing was causing the application export to create duplicate columns. This migration should merge that data together into the new way

## How Can This Be Tested/Reviewed?
okay so testing this is kind of a pain locally, my recommendation is to go as follows:

1) before checking this branch out run a reseed
2) check this branch out
3) insert data into the following tables: 
```
SELECT
    *
FROM programs;

SELECT
    *
FROM preferences;

SELECT
    *
FROM listing_programs;

SELECT
    *
FROM listing_preferences;
```
4) update data in the following table/colums:
```
SELECT
    id,
    programs,
    preferences
FROM applications;
```
The data that you update into here should have data that is keyed off the program/preference instead of the multiselect question

5) run `yarn db:migration:run`

what you should get is the data you added in gets merged into the new way, and thus the extra columns in the export no longer show!

## Checklist:
- [x] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [x] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

**Comment by emilyjablonski** [comment]
👋 Do you mind providing an example of the kinds of data that you're inserting?

File: backend/core/src/migration/1661805250707-programAndPreferenceUpdates.ts [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+              ...selection,
+              key: metaKey.text,
+              options: selection.options.map((opt) => {
+                const toReturn = {
+                  ...opt,
+                }
+                toReturn.key = this.getTranslated(
+                  `${type}s`,
+                  selection.key,
+                  "preferNotToSay" ? "preferNotToSay" : `${opt.key}.label`,
```
Comment (ludtkemorgan) [inline backend/core/src/migration/1661805250707-programAndPreferenceUpdates.ts:262]: I think you meant to do 
```
opt.key === "preferNotToSay" ? "preferNotToSay" : `${option.key}.label`,

**Comment by emilyjablonski** [comment]
@YazeedLoonat Question for you above on what kinds of data can be used to test :)

File: backend/core/src/migration/1661805250707-programAndPreferenceUpdates.ts [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    await queryRunner.query(`
+        INSERT INTO multiselect_programs_preferences_mapper (multiselect_id, program_or_preference_id, application_section, juris_name)
+        SELECT
+            mq.id,
+            p.id,
+            'programs',
+            (
+                SELECT
+                    j.name
+                FROM jurisdictions_programs_programs jpp
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend/core/src/migration/1661805250707-programAndPreferenceUpdates.ts:96]: yeah I think thats how typeorm set up the table names, I think we should seriously consider going through and imposing reasonable names on some of these tables. There are also some points where column names can be wonky:
"listings_id" and "listing_id" both being used to point to a listing's id 
same for "applications_id" and "application_id" 

overall it makes figuring out how the tables are linked together and stuff unnecessarily complicated

**Comment by seanmalbert** [comment]
> 👋 Do you mind providing an example of the kinds of data that you're inserting?

@emilyjablonski , (@YazeedLoonat , correct me if you have a better way in mind) I think the most sure way to test this is to restore a backup of hba production locally and run it. I created another manual backup (b028) for this purpose.

**Comment by YazeedLoonat** [comment]
oof, thanks sean I was trying to create a set of these in a query for people to just run, but using a backup of live catches a bunch more edge cases. 

To add on once you have the back up set up on your local (I used https://sunglee05.medium.com/import-export-database-postgres-to-from-heroku-7e3ea4e325d8 as the walk through minus the part about creating another backup) 

run the new migration with `yarn db:migration: run` 

then its time to compare data, the one I used was listing id: `b5d8e69b-c992-4976-b0fe-a56a7ed93d54` 
what you are looking for is that the data in the `programs` and `preferences` column of the applications table is no longer keyed off of the program/preference key, but instead the new multiselect question text field

**Comment by YazeedLoonat** [comment]
@ludtkemorgan this has also been updated with what you caught

**Comment by emilyjablonski** [comment]
@YazeedLoonat I could be testing this incorrectly, but I saw no change to the columns, and here's what I did: (1) go to dev, reseed (2) download and standup prod backup (3) run csv export on dev (4) come to this branch and run the migration (5) run the csv export again. I used Telegraph Gardens which is already displaying the bug. I expected to see the duplicate preference and programs columns merged, but I didn't see any changes.

**Comment by YazeedLoonat** [comment]
Hey @emilyjablonski I'm not seeing something similar to what you're describing

for Telegraph Gardens all their program data should be removed, since they don't have a program/preference tied to the listing at all

**Comment by ludtkemorgan** [comment]
I attempted to use the downloaded version hba prod. Got the data imported but then running the migrations I get this error
```
query failed:
        CREATE TABLE "multiselect_programs_preferences_mapper" (
            "multiselect_id" uuid NOT NULL,
            "program_or_preference_id" uuid NOT NULL,
            "application_section" "public"."multiselect_questions_application_section_enum" NOT NULL,
            "juris_name" text not null,
            "created_at" TIMESTAMP NOT NULL DEFAULT now()
        )
error: error: relation "multiselect_programs_preferences_mapper" already exists
```
However I was able to test it with some poorly made mock data and it seemed to have worked

**Review by ludtkemorgan — APPROVED** [review]
I think I feel confident enough to approve this and we can do some testing on stage with a cloned db of prod

**Comment by emilyjablonski** [comment]
Note: From an offline review, splitting preference options into each having their own column is coming in a separate PR

**Comment by YazeedLoonat** [comment]
When qa'ing please look for missing translations

**Review by seanmalbert — APPROVED** [review]
With the update you're pushing up, this looks good!

**Review by emilyjablonski — APPROVED** [review]
Do you mind making the new issue for second piece here of splitting options into separate columns to add whatever context is needed? I think that piece is wanted soon as well

**Comment by YazeedLoonat** [comment]
Hey @emilyjablonski I spun up https://github.com/bloom-housing/bloom/issues/3045 per your note

**Pull request merged**